solr/src/main/java/org/apache/metamodel/solr/SolrDataContext.java

```diff
@@ -0,0 +1,429 @@
+	private final String indexName;
+	private final String url;
+
+	private SolrQuery query = new SolrQuery();
+
+	private static enum queryTypes {
+		ROW, FACET
+	};
+
+	private static final int MAX_RETRIES = 0;
+	private static final int SOCK_TIMEOUT = 1000;
+	private static final int CONN_TIMEOUT = 5000;
+	private static final int DEFAULT_LIMIT = 1000;
+
+	public SolrDataContext(String url, String indexName) {
+		this.url = url;
+		this.indexName = indexName;
+		this.tableDef = detectSchema(url, indexName);
+	}
+
+	private SimpleTableDef detectSchema(String url, String indexName) {
+		SimpleTableDef tableDef = null;
+
+		url += "/schema";
+
+		try {
+			HttpClient httpclient = new DefaultHttpClient();
+			HttpGet request = new HttpGet(url);
+			HttpResponse response = httpclient.execute(request);
+
+			InputStreamReader inputStream = new InputStreamReader(response
+					.getEntity().getContent());
+
+			ObjectReader reader = new ObjectMapper().reader(HashMap.class);
+			Map<String, Object> outerMap = reader.readValue(inputStream);
+
+			Map<String, Object> schemaMap = (Map) outerMap.get("schema");
+
+			if (outerMap != null) {
+				List<Map<String, String>> fieldMapList = (List) schemaMap
+						.get("fields");
+				List<String> columnNames = new ArrayList<String>();
+				List<ColumnType> columnTypes = new ArrayList<ColumnType>();
+
+				for (int i = 0; i < fieldMapList.size(); i++) {
+					Map<String, String> fieldMap = fieldMapList.get(i);
+
+					columnNames.add(fieldMap.get("name"));
+					columnTypes.add(ColumnType.VARCHAR);
+				}
+
+				String[] columnNamesArr = new String[columnNames.size()];
+				columnNamesArr = columnNames.toArray(columnNamesArr);
+
+				ColumnType[] columnTypesArr = new ColumnType[columnTypes.size()];
+				columnTypesArr = columnTypes.toArray(columnTypesArr);
+
+				tableDef = new SimpleTableDef(indexName, columnNamesArr,
+						columnTypesArr);
+			}
+		} catch (IOException e) {
+			logger.error("Failed to parse schema", "", e);
+		}
+
+		return tableDef;
+	}
+
+	@Override
+	protected Schema getMainSchema() throws MetaModelException {
+		try {
+			final MutableSchema theSchema = new MutableSchema(
+					getMainSchemaName());
+
+			final MutableTable table = tableDef.toTable().setSchema(theSchema);
+			final Column idColumn = table.getColumnByName(FIELD_ID);
+			if (idColumn != null && idColumn instanceof MutableColumn) {
+				final MutableColumn mutableColumn = (MutableColumn) idColumn;
+				mutableColumn.setPrimaryKey(true);
+			}
+			theSchema.addTable(table);
+
+			return theSchema;
+		} catch (Exception e) {
+			throw new MetaModelException("Schema retrieval failed " + e);
+		}
+	}
+
+	@Override
+	protected Number executeCountQuery(Table table,
+			List<FilterItem> whereItems, boolean functionApproximationAllowed) {
+		String queryStr = "*:*";
+
+		if (whereItems != null) {
+			for (int i = 0; i < whereItems.size(); i++) {
+				FilterItem whereItem = whereItems.get(i);
+				queryStr = whereItem.toString();
```

Comment:
the 'queryStr' variable is being overridden for each time in the loop. So if more WHERE items are there, only the last will be effective.

I also am doubtful if Solr's query language is similar to the toString() on FilterItem. Does this work?